editor/editor_settings_dialog.cpp

```diff
@@ -74,7 +75,82 @@ void EditorSettingsDialog::_settings_property_edited(const String &p_name) {
+		orbit_mod_key_1 = InputEventKey::create_reference(Key::NONE);
+		orbit_mod_key_2 = InputEventKey::create_reference(Key::NONE);
+		pan_mod_key_1 = InputEventKey::create_reference(Key::SHIFT);
+		pan_mod_key_2 = InputEventKey::create_reference(Key::NONE);
+		zoom_mod_key_1 = InputEventKey::create_reference(Key::SHIFT);
+		zoom_mod_key_2 = InputEventKey::create_reference(Key::CTRL);
+	} else if (nav_scheme == Node3DEditorViewport::NAVIGATION_MAYA) {
+		set_preset = true;
+		set_orbit_mouse_button = Node3DEditorViewport::NAVIGATION_LEFT_MOUSE;
+		set_pan_mouse_button = Node3DEditorViewport::NAVIGATION_MIDDLE_MOUSE;
+		set_zoom_mouse_button = Node3DEditorViewport::NAVIGATION_RIGHT_MOUSE;
+		orbit_mod_key_1 = InputEventKey::create_reference(Key::ALT);
+		orbit_mod_key_2 = InputEventKey::create_reference(Key::NONE);
+		pan_mod_key_1 = InputEventKey::create_reference(Key::NONE);
+		pan_mod_key_2 = InputEventKey::create_reference(Key::NONE);
+		zoom_mod_key_1 = InputEventKey::create_reference(Key::ALT);
+		zoom_mod_key_2 = InputEventKey::create_reference(Key::NONE);
+	} else if (nav_scheme == Node3DEditorViewport::NAVIGATION_MODO) {
+		set_preset = true;
+		set_orbit_mouse_button = Node3DEditorViewport::NAVIGATION_LEFT_MOUSE;
+		set_pan_mouse_button = Node3DEditorViewport::NAVIGATION_LEFT_MOUSE;
+		set_zoom_mouse_button = Node3DEditorViewport::NAVIGATION_LEFT_MOUSE;
+		orbit_mod_key_1 = InputEventKey::create_reference(Key::ALT);
+		orbit_mod_key_2 = InputEventKey::create_reference(Key::NONE);
+		pan_mod_key_1 = InputEventKey::create_reference(Key::SHIFT);
+		pan_mod_key_2 = InputEventKey::create_reference(Key::ALT);
+		zoom_mod_key_1 = InputEventKey::create_reference(Key::ALT);
+		zoom_mod_key_2 = InputEventKey::create_reference(Key::CTRL);
+	}
+	// Set settings to the desired preset values.
+	if (set_preset) {
+		EditorSettings::get_singleton()->set_manually("editors/3d/navigation/orbit_mouse_button", (int)set_orbit_mouse_button);
+		EditorSettings::get_singleton()->set_manually("editors/3d/navigation/pan_mouse_button", (int)set_pan_mouse_button);
+		EditorSettings::get_singleton()->set_manually("editors/3d/navigation/zoom_mouse_button", (int)set_zoom_mouse_button);
+		_set_shortcut_input("spatial_editor/viewport_orbit_modifier_1", orbit_mod_key_1);
+		_set_shortcut_input("spatial_editor/viewport_orbit_modifier_2", orbit_mod_key_2);
+		_set_shortcut_input("spatial_editor/viewport_pan_modifier_1", pan_mod_key_1);
+		_set_shortcut_input("spatial_editor/viewport_pan_modifier_2", pan_mod_key_2);
+		_set_shortcut_input("spatial_editor/viewport_zoom_modifier_1", zoom_mod_key_1);
+		_set_shortcut_input("spatial_editor/viewport_zoom_modifier_2", zoom_mod_key_2);
+	}
+}
+
+void EditorSettingsDialog::_set_shortcut_input(const String &p_name, Ref<InputEventKey> &p_event) {
+	Ref<Shortcut> sc = EditorSettings::get_singleton()->get_shortcut(p_name);
+	Array sc_events = sc->get_events();
+	sc_events.clear();
```

Comment:
Why the clear and then reassignment?

Reply:
It seems `set_events()` is necessary for the shortcut to run `emit_changed()` and breaks otherwise. But good catch since we don't need to do the get_events() step, I can just make it a new Array. A question though, do I have to do anything to free up the memory used by the old array when it's replaced or will Godot take care of that? I think I did it this way originally to avoid any complications with that

Reply:
It will, it's fully safe, that's not a problem